File: 18. Backend Scaling and Performance Engineering-part-1/code/typescript/example_4.ts
import { Pool } from 'pg';

export async function getPostsWithAuthors(db: Pool) {
    const postsRes = await db.query('SELECT * FROM posts');
    const posts = postsRes.rows;

    const authorIDSet = new Set(posts.map(p => p.author_id));
    const authorIDs = Array.from(authorIDSet);
    
    if (authorIDs.length === 0) return [];

    const authorsRes = await db.query('SELECT * FROM authors WHERE id = ANY($1)', [authorIDs]);
    const authors = authorsRes.rows;

    const authorMap = new Map();
    authors.forEach(a => authorMap.set(a.id, a));

    return posts.map(p => ({
        post: p,
        author: authorMap.get(p.author_id)
    }));
}

File: 18. Backend Scaling and Performance Engineering-part-1/code/rust/example_4.rs
use sqlx::{PgPool, FromRow};
use std::collections::{HashSet, HashMap};

#[derive(FromRow, Clone)]
pub struct Post { pub id: i64, pub author_id: i64 }
#[derive(FromRow, Clone)]
pub struct Author { pub id: i64 }
pub struct PostWithAuthor { pub post: Post, pub author: Option<Author> }

pub async fn get_posts_with_authors(db: &PgPool) -> Result<Vec<PostWithAuthor>, sqlx::Error> {
    let posts: Vec<Post> = sqlx::query_as("SELECT * FROM posts").fetch_all(db).await?;
    
    let author_ids: Vec<i64> = posts.iter()
        .map(|p| p.author_id)
        .collect::<HashSet<_>>()
        .into_iter()
        .collect();

    let authors: Vec<Author> = sqlx::query_as("SELECT * FROM authors WHERE id = ANY($1)")
        .bind(&author_ids)
        .fetch_all(db)
        .await?;

    let author_map: HashMap<i64, Author> = authors.into_iter().map(|a| (a.id, a)).collect();

    let result = posts.into_iter().map(|post| {
        let author = author_map.get(&post.author_id).cloned();
        PostWithAuthor { post, author }
    }).collect();

    Ok(result)
}
